cookbook/rag_fusion.ipynb

```diff
@@ -0,0 +1,346 @@
+    "//  const prompt = ChatPromptTemplate.fromMessages([\n",
+    "//    [\"system\", \"You are a helpful assistant that generates multiple search queries based on a single input query.\"],\n",
+    "//    [\"user\", \"Generate multiple search queries related to: {original_query}\"],\n",
+    "//    [\"user\", \"OUTPUT (4 queries):\"],\n",
+    "//  ]);"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 66,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "/** Define our chain for generating queries  */\n",
+    "const generateQueries = RunnableSequence.from([\n",
+    "  prompt,\n",
+    "  model,\n",
+    "  new StringOutputParser(),\n",
+    "  RunnableLambda.from((output) => output.split(\"\\n\")),\n",
+    "]);"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Construct the Reciprocal Rank Fusion function\n",
+    "This function is used for combining the results of multiple search queries to produce a single ranked list of results. This is a common technique in information retrieval known as data fusion or result merging."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 67,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "import { Document } from \"npm:langchain@0.0.172/document\";"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 68,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "const reciprocalRankFusion = (results: Document[][], k = 60) => {\n",
+    "  const fusedScores: Record<string, number> = {};\n",
+    "  for (const result of results) {\n",
+    "    // Assumes the docs are returned in sorted order of relevance\n",
+    "    result.forEach((item, index) => {\n",
+    "      const docString = item.pageContent;\n",
+    "      if (!(docString in fusedScores)) {\n",
+    "        fusedScores[docString] = 0;\n",
+    "      }\n",
+    "      fusedScores[docString] += 1 / (index + k);\n",
+    "    });\n",
+    "  }\n",
+    "\n",
+    "  const rerankedResults = Object.entries(fusedScores)\n",
+    "    .sort((a, b) => b[1] - a[1])\n",
+    "    .map(\n",
+    "      ([doc, score]) => new Document({ pageContent: doc, metadata: { score } })\n",
+    "    );\n",
+    "  return rerankedResults;\n",
+    "};"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Define the full chain\n",
+    "Now we can put all our pieces together in one chain.\n",
+    "The chain preforms the following steps:\n",
+    "1. Generate 4 search queries based on the original query\n",
+    "2. Perform lookups with the retriever for each generated query\n",
+    "3. Pass the results of the vector store lookup to the `reciprocalRankFusion` function"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 69,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "const chain = RunnableSequence.from([\n",
+    "  generateQueries,\n",
+    "  retriever.map(),\n",
```

Comment:
Ah I forgot this method existed 😅

Reply:
Can you make a note to document in the LCEL section?

Reply:
tracking here https://github.com/langchain-ai/langchainjs/issues/3052